### PR TITLE
Benchmark on_runtime_upgrade

### PR DESCRIPTION
**Pull request by ferrell-code**
relates #10064 

Basic attempt to benchmark on_runtime_upgrade with try--runtime, any feedback is welcome as I'm not sure how accurate of a benchmark this is

### CODE REVIEW

File: utils/frame/try-runtime/cli/src/commands/benchmark_upgrade.rs [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+	command: BenchmarkUpgradeCmd,
+	config: Configuration,
+) -> sc_cli::Result<()>
+where
+	Block: BlockT + serde::de::DeserializeOwned,
+	Block::Hash: FromStr,
+	<Block::Hash as FromStr>::Err: Debug,
+	NumberFor<Block>: FromStr,
+	<NumberFor<Block> as FromStr>::Err: Debug,
+	ExecDispatch: NativeExecutionDispatch + 'static,
```
Comment (thiolliere) [inline utils/frame/try-runtime/cli/src/commands/benchmark_upgrade.rs:53]: so that mean we are benchmarking the native execution ? not the wasm one ?

**Review by thiolliere — COMMENTED** [review]
another direction would be to add the benchmarking of the custom runtime upgrade inside the benchmark of the runtime. But it doesn't seem easy

it would look like this. So that OnRuntimeUpgrade has a correct weight.

```diff
diff --git a/bin/node/runtime/src/lib.rs b/bin/node/runtime/src/lib.rs
index ce53a3a796..8dea07295f 100644
--- a/bin/node/runtime/src/lib.rs
+++ b/bin/node/runtime/src/lib.rs
@@ -1328,9 +1328,19 @@ pub type Executive = frame_executive::Executive<
        frame_system::ChainContext<Runtime>,
        Runtime,
        AllPallets,
-       (),
+       CustomOnRuntimeUpgrade,
 >;
 
+mod CustomOnRuntimeUpgradeWeightInfo;
+
+pub struct CustomOnRuntimeUpgrade;
+impl OnRuntimeUpgrade for CustomOnRuntimeUpgrade {
+       fn on_runtime_upgrade(..) -> Weight {
+               ....
+               CustomOnRUntimeWeightInfo::weight()
+       }
+}
+
 /// MMR helper types.
 mod mmr {
        use super::Runtime;
@@ -1731,6 +1741,8 @@ impl_runtime_apis! {
                        add_benchmark!(params, batches, pallet_uniques, Uniques);
                        add_benchmark!(params, batches, pallet_utility, Utility);
                        add_benchmark!(params, batches, pallet_vesting, Vesting);
+                       // TODO: add the benchmark for on_runtime_ugprade
+                       // But it is not straightforward
 
                        Ok(batches)
                }
```

but your way may be fine.

Sidenote: also anywhere when we do return weight being max_block in on_initialize, it will to be fixed for parachain, to ensure to doesn't exhaust the block weight

File: utils/frame/try-runtime/cli/src/commands/benchmark_upgrade.rs (hunk 1 above)
Comment (ferrell-code) [reply utils/frame/try-runtime/cli/src/commands/benchmark_upgrade.rs:53]: No the executor is `NativeElseWasmExecutor`, giving the option for either. So in my example output used `--execution wasm`. But that is a good point I should make it so native execution is not an option

**Comment by ferrell-code** [comment]
Yeah I agree your suggestion is much cleaner and fits with how benchmarking is done in the rest of substrate, although as you pointed out it is much more difficult. Also I think this pr is incorrect as caching makes it so the loop of `on_runtime_upgrade` are faster than reality making the calculated weight wrong. I think I will close it as I don't like the code

>Sidenote: also anywhere when we do return weight being max_block in on_initialize, it will to be fixed for parachain, to ensure to doesn't exhaust the block weight

thank you, i figured that out after a bit of confusion :p

**Pull request closed without merging**